File: leet_code/linked_lists/detect_cycle.rb
def detect_cycle(head)
  return nil if !head || !head.next

  slow = head
  fast = head

  while fast && fast.next
    slow = slow.next
    fast = fast.next.next
    break if fast == slow
  end

  return nil if !fast || !fast.next

  fast = head
  until fast == slow
    slow = slow.next
    fast = fast.next
  end
  fast
end

File: leet_code/linked_lists/detect_cycle.js
const detectCycle = head => {
  if (!head || !head.next) return null;

    let slow = head;
    let fast = head;

    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow == fast) break;
    }

    if (!fast || !fast.next) return null;

    fast = head;
    while (fast != slow) {
        slow = slow.next;
        fast = fast.next;
    }
    return fast;
};
